README.md
<p align="center">
	<img align="center" src="https://raw.githubusercontent.com/JOTSR/Deplot/main/public/logo.png"  />
	<br>
</p>

# Deplot

Simple and complete Plot gui module for [Deno](https://deno.land) local scripts
to provide a helper for science computing.

[![deno module](https://shield.deno.dev/x/deplot)](https://deno.land/x/deplot)
![deno compatibility](https://shield.deno.dev/deno/^1.30)
[![vr scripts](https://badges.velociraptor.run/flat.svg)](https://velociraptor.run)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/deplot/mod.ts)

## Dependencies

Use [WebUI](https://deno.land/x/webui) for window managing and
plots libraries for tracing.

Supports [Plotly.js](https://plotly.com/javascript/) and [Chart.js](https://www.chartjs.org/) for plotting.

## Usage

All the modules are exposed in `mod.ts`

```ts
import { Datas, Deplot, Plotly } from 'https://deno.land/x/deplot/mod.ts'

const barPlot = new Deplot(
	'Plotly',
	{
		title: 'My bar plot',
		size: { width: 800, height: 600 },
	},
)

const trace: Plotly.Data = {
	x: ['Zebras', 'Lions', 'Pelicans'],
	y: [90, 40, 60],
	type: 'bar',
	name: 'New York Zoo',
}

const layout: Partial<Plotly.Layout> = {
	title: 'Hide the Modebar',
	showlegend: true,
}

const datas: Datas = { data: [trace], layout, config: { editable: true } }

await barPlot.plot(datas)

const trace2: Plotly.Data = {
	x: [1, 2, 3, 4, 5],
	y: [4, 0, 4, 6, 8],
	mode: 'lines+markers',
	type: 'scatter',
}

const datas2 = { data: [trace2], layout }

new Deplot('Plotly').plot(datas)
```

[Documentation](https://doc.deno.land/https://deno.land/x/deplot/mod.ts)

## Examples

[Plotly.js](https://plotly.com/javascript/)

```sh
deno run --allow-read --allow-write --allow-env --allow-ffi --unstable https://deno.land/x/deplot/examples/plotly.ts
```

[Chart.js](https://www.chartjs.org/docs/3.7.0/)

```sh
deno run --allow-read --allow-write --allow-env --allow-ffi --unstable https://deno.land/x/deplot/examples/chartjs.ts
```

_For flags see [Deno WebUI/Security flags](https://github.com/webui-dev/deno-webui/#security-flags)_


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [1.1.0](https://github.com/JOTSR/Deplot/compare/1.0.0...1.1.0)

> 25 July 2023

- feat: :sparkles: add dark/light mode configuration [`66ede72`](https://github.com/JOTSR/Deplot/commit/66ede72b4d62b1c1a6e2d7314bac535d492cc711)
- docs: :memo: add changelog [`624cfb4`](https://github.com/JOTSR/Deplot/commit/624cfb4482b48879e5a2f04ee08f7eb792df2435)

### [1.0.0](https://github.com/JOTSR/Deplot/compare/0.3.0...1.0.0)

> 21 July 2023

- refactor: :recycle: move Deplot to WebUI for UI rendering [`#1`](https://github.com/JOTSR/Deplot/issues/1)
- style: :art: update fmt without semicolons [`b172746`](https://github.com/JOTSR/Deplot/commit/b1727465a49130bba8a4cb51fcd0a24910455760)
- refactor: :recycle: move all imports to deps.ts and sort imports [`bc8361b`](https://github.com/JOTSR/Deplot/commit/bc8361b010b5d34a54ff03a0658b7f15f711b6b2)
- refactor: :recycle: split plot datas and window options [`aeb8133`](https://github.com/JOTSR/Deplot/commit/aeb81334bf5b262eda142ff736569b739b699198)
- refactor: :recycle: rewrite client to new UI based API [`b81dc48`](https://github.com/JOTSR/Deplot/commit/b81dc48e95e80bf8de9b9867022f7f7f15c96c44)
- docs: :memo: improve and add examples to JSDoc [`bb0b67d`](https://github.com/JOTSR/Deplot/commit/bb0b67d89961f1090b3c8ce19559a6b4716cad4b)
- docs: :memo: update module documentation [`174b1a9`](https://github.com/JOTSR/Deplot/commit/174b1a96c881aa4cefaa63a783944ffdf1d32685)
- chore: update deno lock file [`6aa49d4`](https://github.com/JOTSR/Deplot/commit/6aa49d492bd1a4cdf47aee12d7e1a2a9955bef12)
- docs: :memo: add Deplot.capture example [`874606a`](https://github.com/JOTSR/Deplot/commit/874606a72ba942da0bcdceeb3f86db4035214263)
- docs: :memo: add JSDoc to public API [`e84ec83`](https://github.com/JOTSR/Deplot/commit/e84ec8372e327c8a8b8aab88d0394dda4adce434)
- build: :construction_worker: rewrite build system and embed favicon to bundle [`510775e`](https://github.com/JOTSR/Deplot/commit/510775e4a3a5d33e0bc73fe372326aa9015342f1)
- feat: :sparkles: implement capture on Deplot [`1ea3307`](https://github.com/JOTSR/Deplot/commit/1ea3307c45659bf16842135201b132c70cec5268)
- refactor: :recycle: rewrite the build system and deprecate deno bundle [`85ecad9`](https://github.com/JOTSR/Deplot/commit/85ecad9b7af1813eaa8c440aa60530a94a71136e)
- docs: :memo: update examples to new WebUI based API [`06e053b`](https://github.com/JOTSR/Deplot/commit/06e053b5d19dcb257e5757ccb67e35f033f18eb2)
- refactor: :coffin: remove dead code [`2561a89`](https://github.com/JOTSR/Deplot/commit/2561a89dc268289273a2120250a0f444f2b4b2e2)
- fix: :safety_vest: wrong boolean condition [`55c0eca`](https://github.com/JOTSR/Deplot/commit/55c0eca4ee9a20f2bc8a260b1e8ab83d0d3eb6d9)
- refactor: :recycle: remove unused types [`8bc82ef`](https://github.com/JOTSR/Deplot/commit/8bc82ef4de87f2a7a48ef198f06d8dec39e9d476)
- refactor: reduce code duplication [`bf42532`](https://github.com/JOTSR/Deplot/commit/bf425324584fb57efea52854fd60448d342120c4)
- fix: :bug: deplot can crash when multiple windows are open [`238ed33`](https://github.com/JOTSR/Deplot/commit/238ed33da1fe90f3f0cd0a90833591b5f013acc3)
- docs: :memo: update README [`bf57a42`](https://github.com/JOTSR/Deplot/commit/bf57a42b6f126ce30b67155bef5d372b8afecb2e)
- refactor: :lipstick: improve plot css layout [`b07eeaf`](https://github.com/JOTSR/Deplot/commit/b07eeaf915fe65c2abbe5d7c438a09656706aa94)
- feat: :sparkles: add Deplot.setPosition [`38af74a`](https://github.com/JOTSR/Deplot/commit/38af74acd4eb73c6890b017a73200156ca4616d8)
- feat: :sparkles: implement capture plot for DeplotClient [`9bde856`](https://github.com/JOTSR/Deplot/commit/9bde8566519b1a13567f6911034375bea6ce448d)
- chore: :recycle: flatten deno.json configuration to follow new format [`d7478af`](https://github.com/JOTSR/Deplot/commit/d7478af25771a4f0e4bae41315fb3b59611042fa)
- fix: :bug: Deplot.capture crash or output invalid png files [`295f3e1`](https://github.com/JOTSR/Deplot/commit/295f3e16a7fc8d7234a571e5428e3a07c8b70d5e)
- refactor: :recycle: update API don't return instance in each methods [`13485a9`](https://github.com/JOTSR/Deplot/commit/13485a96967579692e5f2e0edb96d0e9f3517135)
- feat: :label: improve experience by using generics types [`ed33f40`](https://github.com/JOTSR/Deplot/commit/ed33f4044e96fcd7c8890937e100f922f415e252)
- feat: :sparkles: expose isShown WebUI API to Deplot instance [`1b15ac3`](https://github.com/JOTSR/Deplot/commit/1b15ac3f0e596653deb4235613bd64856279a2a3)
- refactor: :fire: drop GChart support [`6f69798`](https://github.com/JOTSR/Deplot/commit/6f69798fcb872a6d5a20163770e806fb8cd80f5a)
- fix: :bug: force Chartjs render and await animation ends to capture plot [`326de00`](https://github.com/JOTSR/Deplot/commit/326de007d4b3d1ce657e16041db93222fea8e92a)
- refactor: :arrow_up: update WebUI version [`7d5d061`](https://github.com/JOTSR/Deplot/commit/7d5d0612cdb688529b2251ff5b673933eae4903b)
- fix: :bug: size not correctly set in Deplot constructor [`870e53c`](https://github.com/JOTSR/Deplot/commit/870e53cba0ec4fc6f631d971090722d09acb94a5)
- refactor: :fire: drop closeCallback support until working solution [`421314f`](https://github.com/JOTSR/Deplot/commit/421314f025b75541ed6cb1c86427d30933c4119a)
- fix: patch PlotlyJs import and vendoring [`45adb85`](https://github.com/JOTSR/Deplot/commit/45adb85a7c1d30653ef4294c089ad32a6f83da79)
- fix: :label: fix Deplot contructor type [`03a9d5f`](https://github.com/JOTSR/Deplot/commit/03a9d5f958b751626d0cfc0bb70a17bd0bdc0039)
- chore: :arrow_up: update deno/std dependencies version [`5491020`](https://github.com/JOTSR/Deplot/commit/549102091ebf4eb23eca354c36d1a202fdcbf72b)
- refactor: :recycle: export client from module to ensure fix name [`04c95a0`](https://github.com/JOTSR/Deplot/commit/04c95a0db4cabd3a2bb3c9ba6e5b8de318c45cd2)

mod.ts
/**
 * # Deplot
 *
 * > Simple and complete Plot gui module for [Deno](https://deno.land) local scripts
 * > to provide a helper for science computing.
 *
 * ## Dependencies
 *
 * Use [WebUI](https://deno.land/x/webui) for window managing and plots libraries for tracing.
 *
 * Supports [Plotly.js](https://plotly.com/javascript/) and [Chart.js](https://www.chartjs.org/) for plotting.
 *
 * ## Usage
 *
 * All the modules are exposed in `mod.ts`
 *
 * ```ts
 * import { Datas, Deplot, Plotly } from 'https://deno.land/x/deplot/mod.ts'
 *
 * const barPlot = new Deplot(
 * 	'Plotly',
 * 	{
 * 		title: 'My bar plot',
 * 		size: { width: 800, height: 600 },
 * 	},
 * )
 *
 * const trace: Plotly.Data = {
 * 	x: ['Zebras', 'Lions', 'Pelicans'],
 * 	y: [90, 40, 60],
 * 	type: 'bar',
 * 	name: 'New York Zoo',
 * }
 *
 * const layout: Partial<Plotly.Layout> = {
 * 	title: 'Hide the Modebar',
 * 	showlegend: true,
 * }
 *
 * const datas: Datas = { data: [trace], layout, config: { editable: true } }
 *
 * await barPlot.plot(datas)
 *
 * const trace2: Plotly.Data = {
 * 	x: [1, 2, 3, 4, 5],
 * 	y: [4, 0, 4, 6, 8],
 * 	mode: 'lines+markers',
 * 	type: 'scatter',
 * }
 *
 * const datas2 = { data: [trace2], layout }
 *
 * new Deplot('Plotly').plot(datas)
 * ```
 *
 * ## Examples
 *
 * [Plotly.js](https://plotly.com/javascript/)
 *
 * ```sh
 * deno run --allow-read --allow-write --allow-env --allow-ffi --unstable https://deno.land/x/deplot/examples/plotly.ts
 * ```
 *
 * [Chart.js](https://www.chartjs.org/docs/3.7.0/)
 *
 * ```sh
 * deno run --allow-read --allow-write --allow-env --allow-ffi --unstable https://deno.land/x/deplot/examples/chartjs.ts
 * ```
 *
 * _For flags see [Deno WebUI/Security flags](https://github.com/webui-dev/deno-webui/#security-flags)_
 *
 * @module
 * @license MIT
 */
import { Deplot } from './src/deplot.ts'
export default Deplot

export { Deplot } from './src/deplot.ts'

export type {
	ChartJs,
	Datas,
	DeplotOptions,
	PlotEngine,
	Plotly,
} from './src/types.ts'
